core/sequencer/beam_fns.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sequencer
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/google/keytransparency/core/mutator/entry"
+
+	ktpb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	spb "github.com/google/keytransparency/core/sequencer/sequencer_go_proto"
+	tpb "github.com/google/trillian"
+	tclient "github.com/google/trillian/client"
+)
+
+// splitMeta emits one KV<logID, source> for each source in the directoryID/Revision batch.
+func splitMeta(meta *spb.MapMetadata, emit func(logID int64, source *spb.MapMetadata_SourceSlice)) {
+	for logID, source := range meta.Sources {
+		glog.Infof("emit(logID %v, source %v)", logID, source)
+		emit(logID, source)
+	}
+}
+
+// ReadLog reads from source.LowestWatermark to source.HighestWatermark in
+// increments of batchSize, emitting *ktpb.EntryUpdates as it goes.
+func (s *Server) ReadLog(ctx context.Context, logID int64, source *spb.MapMetadata_SourceSlice,
```

Comment:
observation: You only need `s.logs`, not sure if it makes sense to make is `Server`'s method rather than passing `s.logs` in directly.
concern: `ReadLog` is supposed to be serialized and distributed among the pipeline workers, isn't it? Will it be able to distribute the `Server` object and/or `s.logs`?